### PR TITLE
Combine spill strategies

### PR DESCRIPTION
**Pull request by rschlussel**
Test plan - unit tests

```
== RELEASE NOTES ==

General Changes
* Remove spilling strategy ``PER_QUERY_MEMORY__LIMIT`` and instead add configuration property ``experimental.query-limit-spill-enabled`` and session property ``query_limit_spill_enabled``.  When this property is set to ``true``, and the spill strategy is not ``PER_TASK_MEMORY_THRESHOLD``, then we will spill whenever a query uses more than the per-node total memory limit in combined revocable and non-revocable memory, in addition to whenever the memory pool exceeds the spill threshold.  This fixes an issue where using the ``PER_QUERY_MEMORY_LIMIT`` spilling strategy could prevent the oom killer from running when the memory pool was full.  The issue is still present for the ``PER_TASK_MEMORY_THRESHOLD`` spilling strategy.

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 1]
```diff
@@ -35,73 +36,89 @@
...
             SqlTaskManager sqlTaskManager,
-            TaskManagementExecutor taskManagementExecutor,
             FeaturesConfig config)
     {
         this(
                 ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
                 requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getAllTasks,
-                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor(),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                (ThreadPoolExecutor) newFixedThreadPool(1, threadsNamed("memory-revocation")),
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java:88]: Yeah, it was entirely to be able to get the corePoolCount. I'll change it.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 2]
```diff
@@ -35,73 +36,89 @@
...
 {
     private static final Logger log = Logger.get(MemoryRevokingScheduler.class);
 
     private static final Ordering<SqlTask> ORDER_BY_CREATE_TIME = Ordering.natural().onResultOf(SqlTask::getTaskCreatedTime);
 
-    private final List<MemoryPool> memoryPools;
+    private final Function<QueryId, QueryContext> queryContextSupplier;
     private final Supplier<List<SqlTask>> currentTasksSupplier;
-    private final ScheduledExecutorService taskManagementExecutor;
+    private final ExecutorService memoryRevocationExecutor;
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java:68]: It wasn't created in the constructor for the tests, but I'll actually just change the tests to get the executor from here, and that will remove any need to do any checking about the pool size.

**Comment by rschlussel** [comment]
@pettyjamesm comments addressed and all tests passing

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 3]
```diff
@@ -166,13 +171,20 @@ public void testScheduleMemoryRevoking()
         OperatorContext operatorContext5 = driverContext211.addOperatorContext(5, new PlanNodeId("na"), "na");
 
         List<SqlTask> tasks = ImmutableList.of(sqlTask1, sqlTask2);
-        MemoryRevokingScheduler scheduler = new MemoryRevokingScheduler(singletonList(memoryPool), () -> tasks, executor, 1.0, 1.0, ORDER_BY_CREATE_TIME);
-
+        MemoryRevokingScheduler scheduler = new MemoryRevokingScheduler(
```
Comment (pettyjamesm) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java:174]: Missing `MemoryRevokingScheduler#stop()` which would now leak the single threaded executor until GC finalization. Minor, but worth avoiding. Maybe a `withMemoryRevokingScheduler(<params>, Consumer<MemoryRevokingScheduler>)` helper could help with the boiler-plate of `registerPoolListeners()` and `stop()` but would be a problem with the way that `InterruptedException` is thrown from `awaitAsynchronousCallbacksRun()`...

Open to your thoughts about the best way to resolve the tension between test sanity and ensuring the revoking executor remains single threaded even with future refactoring changes.

Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java:174]: ended up going with a simple try/finally pattern in the tests to start and stop the memory revoking scheduler

**Review by pettyjamesm — APPROVED** [review]

**Pull request merged**